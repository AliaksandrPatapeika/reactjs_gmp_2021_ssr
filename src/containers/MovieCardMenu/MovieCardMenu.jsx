import './MovieCardMenu.less';

import PropTypes from 'prop-types';
import React, {useCallback, useEffect, useRef} from 'react';

import Button from '../../components/Button';

const MovieCardMenu = ({
  showSubMenu, closeSubMenu, showEditMovieWindow, showDeleteMovieWindow
}) => {
  const movieCardMenuCloseButtonRef = useRef();
  /* https://gist.github.com/pstoica/4323d3e6e37e8a23dd59 */
  const handleBlur = (e) => {
    const {currentTarget} = e;

    // Check the newly focused element in the next tick of the event loop
    setTimeout(() => {
      // Check if the new activeElement is a child of the original container
      if (!currentTarget.contains(document.activeElement)) {
        // Invoke a callback here
        closeSubMenu();
      }
    }, 0);
  };

  const handleKeyDown = useCallback(
    (event) => {
      if (event.key === 'Escape') {
        closeSubMenu();
      }
    },
    [closeSubMenu]
  );

  useEffect(() => {
    if (showSubMenu) {
      window.addEventListener('keydown', handleKeyDown);
      movieCardMenuCloseButtonRef.current.focus();
    }
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [showSubMenu, handleKeyDown]);

  return showSubMenu && (
  <div className="movieCardMenuContainer" onBlur={handleBlur}>
    {/* // PATTERN: Avoid Inline Function Definition in the Render Function */}
    <Button className="closeButton" ref={movieCardMenuCloseButtonRef} onClick={closeSubMenu}>
      <i className="fa fa-times" aria-hidden="true" />
    </Button>
    <nav>
      <ul>
        <li
          role="menuitem"
          tabIndex="0"
          className="menuItem"
          // PATTERN: Avoid Inline Function Definition in the Render Function
          onClick={showEditMovieWindow}
          onKeyPress={showEditMovieWindow}
        >
          Edit
        </li>
        <li
          role="menuitem"
          tabIndex="0"
          className="menuItem"
          // PATTERN: Avoid Inline Function Definition in the Render Function
          onClick={showDeleteMovieWindow}
          onKeyPress={showDeleteMovieWindow}
        >
          Delete
        </li>
      </ul>
    </nav>
  </div>
  );
};

MovieCardMenu.propTypes = {
  showSubMenu: PropTypes.bool.isRequired,
  closeSubMenu: PropTypes.func.isRequired,
  showEditMovieWindow: PropTypes.func.isRequired,
  showDeleteMovieWindow: PropTypes.func.isRequired
};

export default MovieCardMenu;
