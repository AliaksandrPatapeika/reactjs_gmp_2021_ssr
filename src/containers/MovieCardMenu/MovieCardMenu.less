@import "../../less/variables";

.movieCardMenuContainer {
	position: absolute;
	right: 8px;
	top: 8px;
	display: flex;
	flex-direction: column;
	width: 160px;
	border-radius: 4px;
	background-color: @black;

	li:last-of-type {
		margin-bottom: 10px;
	}

	.menuItem {
		cursor: pointer;
		padding: 4px 16px;
		color: @white;
		font-size: 18px;
		font-family: @NetflixSans-Regular;

		&:hover,
		&:focus {
			background-color: @red;
		}
	}
}
