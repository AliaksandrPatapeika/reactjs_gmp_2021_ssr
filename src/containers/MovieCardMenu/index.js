import MovieCardMenu from './MovieCardMenu';

export default MovieCardMenu;
