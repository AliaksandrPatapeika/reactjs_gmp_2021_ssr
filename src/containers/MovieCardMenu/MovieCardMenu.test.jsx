import {noop} from 'lodash';
import React from 'react';
import renderer from 'react-test-renderer';

import MovieCardMenu from './MovieCardMenu';

describe('MovieCardMenu', () => {
  test('renders correctly', () => {
    const movieCardMenu = renderer
      .create(<MovieCardMenu
        showSubMenu
        closeSubMenu={noop}
        showEditMovieWindow={noop}
        showDeleteMovieWindow={noop}
      />)
      .toJSON();

    expect(movieCardMenu).toMatchSnapshot();
  });
});
