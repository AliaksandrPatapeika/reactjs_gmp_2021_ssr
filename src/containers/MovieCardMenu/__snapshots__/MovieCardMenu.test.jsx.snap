// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCardMenu renders correctly 1`] = `
<div
  className="movieCardMenuContainer"
  onBlur={[Function]}
>
  <button
    className="closeButton"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    
    <i
      aria-hidden="true"
      className="fa fa-times"
    />
  </button>
  <nav>
    <ul>
      <li
        className="menuItem"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        Edit
      </li>
      <li
        className="menuItem"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        Delete
      </li>
    </ul>
  </nav>
</div>
`;
